extensions/impl/kafka/source.go

```diff
@@ -0,0 +1,289 @@
+
+	"github.com/lf-edge/ekuiper/contract/v2/api"
+	"github.com/lf-edge/ekuiper/v2/internal/conf"
+	"github.com/lf-edge/ekuiper/v2/pkg/cast"
+	"github.com/lf-edge/ekuiper/v2/pkg/cert"
+	"github.com/lf-edge/ekuiper/v2/pkg/timex"
+)
+
+type KafkaSource struct {
+	reader    *kafkago.Reader
+	offset    int64
+	tlsConfig *tls.Config
+	sc        *kafkaSourceConf
+}
+
+type kafkaSourceConf struct {
+	Topic       string `json:"datasource"`
+	Brokers     string `json:"brokers"`
+	GroupID     string `json:"groupID"`
+	Partition   int    `json:"partition"`
+	MaxAttempts int    `json:"maxAttempts"`
+	MaxBytes    int    `json:"maxBytes"`
+}
+
+func (c *kafkaSourceConf) validate() error {
+	if c.Topic == "" {
+		return fmt.Errorf("kafkaSourceConf topic is required")
+	}
+	if len(c.Brokers) < 1 {
+		return fmt.Errorf("brokers can not be empty")
+	}
+	return nil
+}
+
+func (c *kafkaSourceConf) GetReaderConfig() kafkago.ReaderConfig {
+	return kafkago.ReaderConfig{
+		Brokers:     strings.Split(c.Brokers, ","),
+		GroupID:     c.GroupID,
+		Topic:       c.Topic,
+		Partition:   c.Partition,
+		MaxBytes:    c.MaxBytes,
+		MaxAttempts: c.MaxAttempts,
+	}
+}
+
+func getSourceConf(props map[string]interface{}) (*kafkaSourceConf, error) {
+	c := &kafkaSourceConf{
+		MaxBytes:    1e6,
+		MaxAttempts: 3,
+	}
+	err := cast.MapToStruct(props, c)
+	if err != nil {
+		return nil, fmt.Errorf("read properties %v fail with error: %v", props, err)
+	}
+	return c, nil
+}
+
+func (k *KafkaSource) Provision(ctx api.StreamContext, configs map[string]any) error {
+	kConf, err := getSourceConf(configs)
+	failpoint.Inject("kafkaErr", func(val failpoint.Value) {
+		err = mockKakfaSourceErr(val.(int), castConfErr)
+	})
+	if err != nil {
+		conf.Log.Errorf("kafka source config error: %v", err)
+		return err
+	}
+	if err := kConf.validate(); err != nil {
+		return err
+	}
+	tlsConfig, err := cert.GenTLSConfig(configs, "kafka-source")
+	if err != nil {
+		conf.Log.Errorf("kafka tls conf error: %v", err)
+		return err
+	}
+	saslConf, err := getSaslConf(configs)
+	failpoint.Inject("kafkaErr", func(val failpoint.Value) {
+		err = mockKakfaSourceErr(val.(int), saslConfErr)
+	})
+	if err != nil {
+		conf.Log.Errorf("kafka sasl error: %v", err)
+		return err
+	}
+	if err := saslConf.Validate(); err != nil {
+		conf.Log.Errorf("kafka validate sasl error: %v", err)
+		return err
+	}
+	mechanism, err := saslConf.GetMechanism()
+	failpoint.Inject("kafkaErr", func(val failpoint.Value) {
+		err = mockKakfaSourceErr(val.(int), mechanismErr)
+	})
+	if err != nil {
+		conf.Log.Errorf("kafka sasl mechanism error: %v", err)
+		return err
+	}
+	readerConfig := kConf.GetReaderConfig()
+	conf.Log.Infof("topic: %s, brokers: %v", readerConfig.Topic, readerConfig.Brokers)
+	readerConfig.Dialer = &kafkago.Dialer{
+		Timeout:       10 * time.Second,
+		DualStack:     true,
+		TLS:           tlsConfig,
+		SASLMechanism: mechanism,
+	}
+	reader := kafkago.NewReader(readerConfig)
```

Comment:
Move connection to connect func. Provision is supposed to be light-weight. It will be called in planner. Connect will only be called when the rule is run.